### PR TITLE
Support for QuantumRegister in gates supporing qubits

### LINKED ISSUE #352: Error in Rz gate
There is a bug in this line and the line above it (rz gate):

https://github.com/QISKit/qiskit-sdk-py/blob/4f15c99c1dc4b91abd5d54ccc6b94c8e38ac4997/qiskit/extensions/standard/rz.py#L61

It is replicated in rx and ry as well.

### PR DESCRIPTION
**Pull request by 1ucian0**
This PR adds support to gates which support qubits to also accept full register, as the spec defines.

## Description
The gates that take one or more qubit as args, should also support the full `QuantumRegister` as argument. For example: `circuit.h(self.qr)`. The PR also extends the tests for the added code (and more).

## Motivation and Context
Trying to write a test for #352 (and #362) I ended up falling the rabbit hole of supporting `QuantumRegister`s as arguments in the gates that operate over qubits. I noticed that many of the gates had this feature broken.

The most notable cases are:
 * The type of arguments in `barrier` are extended, in order to behave more similarly to other gates. In addition to the previous `*list`, now QuantumRegister are supported, also in combination with qubits. Also, if None is passed, it applied the barrier to every QuantumRegister of the circuit (this was a TODO in the code). 
 * The specification the semantics of `cx` when is called with register and bits (Fig 2, page 4). This semantic is now supported.
 * The gates `cx_base` and `swap` are analogue to `cx`. 
 * The specification do not define the semantics for `ccx` when called with registers. Therefore, I only included the intuitive case in which all the arguments are registers (with the same length): `ccx q,r,s` => `ccx q[0],r[0],s[0];ccx q[1],r[1],s[1];` when length is 2.
 * The `cswap` case is analogue to `ccx`.

BTW, this PR closes #352.

## How Has This Been Tested?
`make test`

The added tests extend the coverage with respect the existing tests:

file | previous stms | previously covered | new stms | now covered
-- | -- | -- | -- | --
qiskit/extensions/standard/__init__.py | 28  | 100%    | 28 | 100%
qiskit/extensions/standard/barrier.py | 45  | 78%      | 40 | 98% 
qiskit/extensions/standard/ccx.py   | 25 | 100%        | 32 | 100%
qiskit/extensions/standard/ch.py    | 30 | 87%         | 40 | 100%
qiskit/extensions/standard/crz.py   | 32 | 88%         | 42 | 100%
qiskit/extensions/standard/cswap.py | 20  | 100%       | 27 | 100%
qiskit/extensions/standard/cu1.py   | 32 | 88%         | 42 | 100%
qiskit/extensions/standard/cu3.py   | 37  | 89%        | 47 | 100%
qiskit/extensions/standard/cx.py    | 30  | 87%        | 40 | 100%
qiskit/extensions/standard/cxbase.py | 23 | 100%       | 40 | 100%
qiskit/extensions/standard/cy.py    | 30  | 87%        | 40 | 100%
qiskit/extensions/standard/cz.py    | 30  | 87%        | 40 | 100%
qiskit/extensions/standard/gatestools.py | 12 | 0%     | 12 | 0%  
qiskit/extensions/standard/h.py     | 27 | 85%         | 32 | 88% 
qiskit/extensions/standard/header.py| 5  | 100%        | 5  | 100%
qiskit/extensions/standard/iden.py  | 27 | 85%         | 27 | 100%
qiskit/extensions/standard/rx.py    | 29 | 86%         | 29 | 100%
qiskit/extensions/standard/ry.py    | 29 | 86%         | 29 | 100%
qiskit/extensions/standard/rz.py    | 29 | 86%         | 29 | 100%
qiskit/extensions/standard/s.py     | 33 | 88%         | 33 | 100%
qiskit/extensions/standard/swap.py  | 23 | 100%        | 40 | 100%
qiskit/extensions/standard/t.py     | 33 | 88%         | 33 | 100%
qiskit/extensions/standard/u0.py    | 27  | 52%        | 27 | 52%
qiskit/extensions/standard/u1.py    | 29 | 86%         | 29 | 100%
qiskit/extensions/standard/u2.py    | 33 | 88%         | 33 | 100%
qiskit/extensions/standard/u3.py    | 34 | 88%         | 34 | 100%
qiskit/extensions/standard/ubase.py | 27 | 100%     | 34 | 100%
qiskit/extensions/standard/x.py     | 27 | 85%         | 27 | 100%
qiskit/extensions/standard/y.py     | 27 | 85%         | 27 | 100%
qiskit/extensions/standard/z.py     | 27 | 85%         | 27 | 100%

None of the changes brake backward compatibility.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by diego-plan9** [comment]
Can you revise the [test that is failing in Appveyor](https://ci.appveyor.com/project/QISKit/qiskit-sdk-py/build/1.0.537/job/waftucqaexfiv0im/tests)? It seems the order of the parameters for `barrier` is different depending on the os or the version:

```
AssertionError: 'barrier q[0],q[1],q[2],r[0],r[1],r[2],s[0],s[1],s[2];' not found in 'OPENQASM 2.0;\ninclude "qelib1.inc";\nqreg q[3];\nqreg r[3];\nqreg s[3];\ncreg c[3];\nbarrier q[0],q[1],q[2],s[0],s[1],s[2],r[0],r[1],r[2];\n'
```

File: qiskit/extensions/standard/barrier.py [hunk 1]
```diff
@@ -52,29 +51,32 @@ def qasm(self):
...
-        raise ExtensionError("no arguments passed")
+        self._modifiers(circ._attach(Barrier(self.arg, self)))
+
+
+def barrier(self, *args):
+    """Apply barrier to circuit.
+    If args is None, applies to all the qbits.
+    If args is a QuantumRegister, applies to all the qbits in that register.
+    If args is a list of QuantumRegister, applies to all of them."""
+
```
Comment (awcross1) [inline qiskit/extensions/standard/barrier.py]: It looks like args can be a list of QuantumRegister and/or qubit, and the barrier will be applied to all arguments. If that's right, could you add that to the comment?

Comment (1ucian0) [reply qiskit/extensions/standard/barrier.py]: Clarified in d3a9d32. Thanks!

File: qiskit/extensions/standard/swap.py [hunk 2]
```diff
@@ -55,8 +55,20 @@ def swap(self, ctl, tgt):
...
+    if isinstance(ctl, QuantumRegister):
+        instructions = InstructionSet()
+        for j in range(ctl.size):
+            instructions.add(self.swap((ctl, j), tgt))
+        return instructions
+
+    if isinstance(tgt, QuantumRegister):
+        instructions = InstructionSet()
+        for j in range(tgt.size):
+            instructions.add(self.swap(ctl, (tgt, j)))
```
Comment (ajavadia) [inline qiskit/extensions/standard/swap.py]: The `swap` gate should not allow ctrl and tgt of different sizes I think.

I think swaps between 2 individual qubits or between 2 registers of the same size makes sense. But when the size of registers differ, an exception should be raised.

@awcross1 do you agree?

Comment (awcross1) [reply qiskit/extensions/standard/swap.py]: @ajavadia Yes, I missed this. I agree that swap should not allow different sized registers as input. Good catch.

Comment (awcross1) [reply qiskit/extensions/standard/swap.py]: Generally we should be careful when G_{i,j} and G_{i,k} do not commute. Does this happen anywhere else in the PR?

Comment (1ucian0) [reply qiskit/extensions/standard/swap.py]: fixed in 42f6b5a. The valid and invalid cases are listed in the tests. I'm not fully sure what you mean with  "G_{i,j} and G_{i,k} cases".

Comment (awcross1) [reply qiskit/extensions/standard/swap.py]: Thanks @1ucian0 the issue doesn't occur anywhere else. I meant G_{i,j} to denote a two-qubit gate acting on qubits i and j.

File: test/python/test_extensions_standard.py [hunk 3]
```diff
@@ -52,182 +52,424 @@
 from .common import QiskitTestCase
 
 
-class TestStandard(QiskitTestCase):
-    """Standard Extension Test."""
+class StandardExtensionTest(QiskitTestCase):
```
Comment (diego-plan9) [inline test/python/test_extensions_standard.py]: Would it be possible to add human-readable docstrings to the assertions methods in this class, ideally describing the parameters as well?

Comment (1ucian0) [reply test/python/test_extensions_standard.py]: fixed in ff05123

**Comment by diego-plan9** [comment]
Thanks @1ucian0 !

**Pull request merged**